### PR TITLE
KAFKA-14352: Rack-aware consumer partition assignment protocol changes (KIP-881)

### PR DESCRIPTION
**Pull request by rajinisivaram**
Protocol changes for KIP-881 to support rack-aware partition assignment for consumers. Default assignors will be updated in subsequent PRs to use racks to improve locality.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractPartitionAssignor.java [hunk 1]
```diff
@@ -47,23 +51,37 @@ public abstract class AbstractPartitionAssignor implements ConsumerPartitionAssi
     public abstract Map<String, List<TopicPartition>> assign(Map<String, Integer> partitionsPerTopic,
                                                              Map<String, Subscription> subscriptions);
 
+    /**
+     * Default implementation of assignPartitions() that does not include racks. This is only
+     * included to avoid breaking any custom implementation that extends AbstractPartitionAssignor.
+     * Note that this class is internal, but to be safe, we are maintaining compatibility.
+     */
+    public Map<String, List<TopicPartition>> assignPartitions(Map<String, List<PartitionInfo>> partitionsPerTopic,
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractPartitionAssignor.java]: Do we need changes in the file for the patch? They do not seem to be related.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocolTest.java [hunk 2]
```diff
@@ -56,7 +56,7 @@ public void serializeDeserializeSubscriptionAllVersions() {
             new TopicPartition("foo", 0),
             new TopicPartition("bar", 0));
         Subscription subscription = new Subscription(Arrays.asList("foo", "bar"),
-            ByteBuffer.wrap("hello".getBytes()), ownedPartitions, generationId);
+            ByteBuffer.wrap("hello".getBytes()), ownedPartitions, generationId, Optional.empty());
```
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocolTest.java]: It would be great if we could extend this test and possibly others in the file to cover the protocol changes.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 3]
```diff
@@ -245,7 +248,8 @@ protected JoinGroupRequestData.JoinGroupRequestProtocolCollection metadata() {
             Subscription subscription = new Subscription(topics,
                                                          assignor.subscriptionUserData(joinedSubscription),
                                                          subscriptions.assignedPartitionsList(),
-                                                         generation().generationId);
+                                                         generation().generationId,
+                                                         rackId);
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:252]: Is it possible to add a unit test which verifies that `rackId` is correctly sent out in the `JoinGroup` request? I am not sure how feasible is this.

**Review by dajac — COMMENTED** [review]
@rajinisivaram Thanks for the patch. I left a few small comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractPartitionAssignor.java (hunk 1 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractPartitionAssignor.java]: Removed.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocolTest.java (hunk 2 above)
Comment (rajinisivaram) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocolTest.java]: Done.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 3 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:252]: Added a unit test. Also added an integration test for verifying propagation since we haven't yet added a rack-aware assignor.

**Comment by rajinisivaram** [comment]
@dajac Thanks for the review, have addressed the comments.

**Review by dajac — APPROVED** [review]
LGTM, thanks!

**Comment by rajinisivaram** [comment]
@dajac Thanks for the review, merging to trunk. Will also confirm with @ableegoldman if we can merge to 3.4 branch.
